buffer/buffer.go

```diff
@@ -0,0 +1,95 @@
+// Package buffer implements an unbounded buffer for ordering concurrent writes for
+// non-seekable writers. It keeps an internal linked list that keeps the chunks in order
+// and flushes buffered chunks when the smallest expected offset is available.
+package buffer
+
+import (
+	"container/list"
+	"io"
+	"sync"
+)
+
+type fileChunk struct {
+	Start   int64
+	Content []byte
+}
+
+type OrderedWriterAt struct {
+	w       io.Writer
+	written int64
+	list    *list.List
+	mu      *sync.Mutex
+}
+
+func NewOrderedWriterAt(w io.Writer) *OrderedWriterAt {
+	return &OrderedWriterAt{
+		written: 0,
+		list:    list.New(),
+		mu:      &sync.Mutex{},
+		w:       w,
+	}
+}
+
+func (ob *OrderedWriterAt) WriteAt(p []byte, offset int64) (int, error) {
```

Comment:
OrderedWriterAt` is the only type with a receiver, so we can just swap `ob` with `w`.